Allow attributes in the `Event::End` and fix `.error_position()`

Mingun opened the pull request:
This PR allows us to have attributes in the `Event::End` (but they are not accessible via `BytesEnd` directly). As explained in https://github.com/tafia/quick-xml/issues/776#issuecomment-2208078867 we anyway can produce an event which `.name()` does not valid, so it is better ar least keep it useful for some scenarios. When validation API would be added (I'm working on it) we can validate all events and report an error.

Closes #776

When changing parser I noticed, that in one arm we do not set `last_error_offset`. I checked if we really cover that lines -- yes, but we actually does not check anything. `last_error_offset` was checked for `0`, but this is its default value, so was unclear is that code running or not. I improved check and fix the bug that was found due to that.

Mingun committed:
Move `Parser`, `ElementParser` and `PiParser` to the new module `parser`.

Mingun committed:
Sort options alphabetically

Mingun committed:
Move tests for allow_unmatched_ends to its own sub-module

Mingun committed:
Remove incorrect example

Mingun committed:
Allow to have attributes in closing tags (compatibility with the Adobe Flash parser)

Mingun committed:
Do not use `Default` implementation to construct parsers

This is not generic code and it is more clear when we construct them with explicit state

Mingun committed:
Assert syntax errors via pretty_assertions

If test will fail due to another syntax error we get a nice diff

Mingun committed:
Actually check that .error_position() is correct

Check position for zero actually does not check anything because this is start
position before parse. We move position by 1 by using Text event

failures (36):
    syntax::tag::unclosed3::ns_reader::async_tokio
    syntax::tag::unclosed3::ns_reader::borrowed
    syntax::tag::unclosed3::ns_reader::buffered
    syntax::tag::unclosed3::reader::async_tokio
    syntax::tag::unclosed3::reader::borrowed
    syntax::tag::unclosed3::reader::buffered
    syntax::tag::unclosed4::ns_reader::async_tokio
    syntax::tag::unclosed4::ns_reader::borrowed
    syntax::tag::unclosed4::ns_reader::buffered
    syntax::tag::unclosed4::reader::async_tokio
    syntax::tag::unclosed4::reader::borrowed
    syntax::tag::unclosed4::reader::buffered
    syntax::tag::unclosed5::ns_reader::async_tokio
    syntax::tag::unclosed5::ns_reader::borrowed
    syntax::tag::unclosed5::ns_reader::buffered
    syntax::tag::unclosed5::reader::async_tokio
    syntax::tag::unclosed5::reader::borrowed
    syntax::tag::unclosed5::reader::buffered
    syntax::tag::unclosed6::ns_reader::async_tokio
    syntax::tag::unclosed6::ns_reader::borrowed
    syntax::tag::unclosed6::ns_reader::buffered
    syntax::tag::unclosed6::reader::async_tokio
    syntax::tag::unclosed6::reader::borrowed
    syntax::tag::unclosed6::reader::buffered
    syntax::tag::unclosed7::ns_reader::async_tokio
    syntax::tag::unclosed7::ns_reader::borrowed
    syntax::tag::unclosed7::ns_reader::buffered
    syntax::tag::unclosed7::reader::async_tokio
    syntax::tag::unclosed7::reader::borrowed
    syntax::tag::unclosed7::reader::buffered
    syntax::tag::unclosed8::ns_reader::async_tokio
    syntax::tag::unclosed8::ns_reader::borrowed
    syntax::tag::unclosed8::ns_reader::buffered
    syntax::tag::unclosed8::reader::async_tokio
    syntax::tag::unclosed8::reader::borrowed
    syntax::tag::unclosed8::reader::buffered

Mingun committed:
Fix incorrect `.error_position()` when encountering syntax error for open or self-closed tag

dralley commented on src/reader/mod.rs:
```diff
@@ -346,18 +346,28 @@ macro_rules! read_until_close {
+            //   do. This is still invalid tag but it also has no practical meaning
+            //   to parse it in a such way. End tag `</tag attr=">" >`, however, can
+            //   be produced by some real-world parsers, for example, by the parser
+            //   of the Macromedia Flash player.
```
For what it's worth, other "non-validating" parsers like expat still blow up on this

>ExpatError: not well-formed (invalid token): line 1, column 11

Lxml and libxml2 also do

>XMLSyntaxError: expected '>', line 1, column 12

>parserError: xmlParseDoc() failed

I'm on-board with making this possible to support, but our default behavior should probably be throwing an `IllFormedError` just like we do for mismatched tags (and like other parsers do).

>ExpatError: mismatched tag: line 1, column 7

Mingun replied:
I want to evolve library in this direction:
- rename `Reader` to `RawReader` (and `Event` to `RawEvent`)
- make `RawReader` as liberal as possible. It should not make any validation checks
- add a `validate` method to `RawEvent` which will perform all well-formedless and validity checks (I already grab list of all checks from the [specification](https://www.w3.org/TR/xml11), 35 VC + 16 WFC references, actually most of them about DTD). All validation must be run manually by calling `RawEvent::validate`. Only one check cannot be performed in this way: check that [start tag name matches end tag name](https://www.w3.org/TR/xml11/#GIMatch)
- move existing checks (for example, [`IllFormedError::DoubleHyphenInComment`](https://github.com/tafia/quick-xml/blob/22b3e458d193e6937882ba6b643ec8c31a671de5/src/errors.rs#L119C26-L119C27)) to the new `validate` method
- introduce a new `Reader` which will perform validation checks when flag in config is set (and process many other things, like dealing with encodings and expanding entity references as shown in the `custom_entities` example in #766). Add necessary config flags

I haven't thought through all the details yet and don't want to focus on them yet. There will be a separate PR where all this will be considered. It should be pretty soon (this year).

So, given that we are not currently performing most validity checks and I am planning a generic consistent solution for them, do you agree that it is not worth introducing them now?

Mingun replied:
> Also line 358 should be rewritten - XML isn't "produced" by a parser. You can just say that such documents exist in the wild and are tolerated by some parsers such as the one used by Adobe Flash.

Any help are appreciated :smiley: . I feel that my English is no good to writing simple-to-read explanations.

dralley replied:
That all sounds good to me, but we probably shouldn't document it as a "new feature" given that we do want this to eventually fail for users of the `Reader` API.

Mingun committed:
Use `.error_position()` instead of `.buffer_position()` in examples where error position is reported

`.error_position()` is intended to report position where error start occurring,
`.buffer_position()` is intended to show to what position reader read data

tests/issues.rs leave untouched because this is the code from real problems in GH issues,
it should remain as close to issue as possible

dralley commented on src/reader/mod.rs:
```diff
@@ -345,18 +345,27 @@ macro_rules! read_until_close {
+            // #776: We parse using ElementParser which allows us to have attributes
+            // in close tags. While such tags are not allowed by the specification,
+            // we anyway allow to parse them because:
+            // - we do not check constrains during parsing. This is performed by the
```
spelling: constraints

dralley commented on src/reader/mod.rs:
```diff
@@ -345,18 +345,27 @@ macro_rules! read_until_close {
+            //   optional validate step which user should call manually
+            // - if we just look for `>` we will parse `</tag attr=">" >` as end tag
+            //   `</tag attr=">` and text `" >` which probably no one existing parser
+            //   do. This is still invalid tag but it also has no practical meaning
```
do -> does

Mingun replied:
Fixed!

Mingun commented:
Task CIFuzz was failed by unknown reasons -- I asked https://github.com/google/oss-fuzz/issues/12168 but no response yet... It not related to this changes, so merge
